File: golang/greed/gas_station.go
package main

import "fmt"

func canCompleteCircuit(gas []int, cost []int) int {
	var remains, total, start int

	for i, g := range gas {
		remains += g - cost[i]
		if remains < 0 {
			start = i + 1
			total += remains
			remains = 0

		}

	}
	if total+remains < 0 {
		return -1
	}
	return start
}
func main() {
	var gas []int = []int{2, 3, 1}
	var cost []int = []int{3, 1, 2}
	fmt.Println(canCompleteCircuit(gas, cost))
}

File: python/gas_station.py
class Solution(object):
    def canCompleteCircuit(self, gas, cost):
        """
        :type gas: List[int]
        :type cost: List[int]
        :rtype: int
        """
        if sum(gas) < sum(cost):
            return -1

        length = len(gas)
        sorted_gas_index = sorted(range(len(gas)), key=lambda k: gas[k])

        for i in sorted_gas_index:
            j = 0
            remain = 0
            while j < length:
                index = (i + j) % length
                if remain + gas[index] < cost[index]:
                    break
                remain += gas[index] - cost[index]
                j += 1
            if j == length:
                return i
        return -1
